src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+                        errors
+                    };
+                }
+                return { scope, scopeDescription: getDescriptionForScope(scope) };
+            });
+        }
+    }
+
+    function getDescriptionForScope(scope: Scope) {
+        if (isFunctionLike(scope)) {
+            switch (scope.kind) {
+                case SyntaxKind.Constructor:
+                    return "constructor";
+                case SyntaxKind.FunctionExpression:
+                    return scope.name
+                        ? `function expression ${scope.name.getText()}`
+                        : "anonymous function expression";
+                case SyntaxKind.FunctionDeclaration:
+                    return `function ${scope.name.getText()}`;
+                case SyntaxKind.ArrowFunction:
+                    return "arrow function";
+                case SyntaxKind.MethodDeclaration:
+                    return `method ${scope.name.getText()}`;
+                case SyntaxKind.GetAccessor:
+                    return `get ${scope.name.getText()}`;
+                case SyntaxKind.SetAccessor:
+                    return `set ${scope.name.getText()}`;
+            }
+        }
+        else if (isModuleBlock(scope)) {
+            return `namespace ${scope.parent.name.getText()}`;
+        }
+        else if (isClassLike(scope)) {
+            return scope.kind === SyntaxKind.ClassDeclaration
+                ? `class ${scope.name.text}`
+                : scope.name.text
+                    ? `class expression ${scope.name.text}`
+                    : "anonymous class expression";
+        }
+        else if (isSourceFile(scope)) {
+            return `file '${scope.fileName}'`;
+        }
+        else {
+            return "unknown";
+        }
+    }
+
+    function getUniqueName(isNameOkay: (name: __String) => boolean) {
+        let functionNameText = "newFunction" as __String;
+        if (isNameOkay(functionNameText)) {
+            return functionNameText;
+        }
+        let i = 1;
+        while (!isNameOkay(functionNameText = `newFunction_${i}` as __String)) {
+            i++;
+        }
+        return functionNameText;
+    }
+
+    export function extractFunctionInScope(
+        node: Statement | Expression | Block,
+        scope: Scope,
+        { usages: usagesInScope, substitutions }: ScopeUsages,
+        range: TargetRange,
+        context: RefactorContext): ExtractResultForScope {
+
+        const checker = context.program.getTypeChecker();
+
+        // Make a unique name for the extracted function
+        let functionNameText: __String;
+        if (isClassLike(scope)) {
+            const type = range.facts & RangeFacts.InStaticRegion ? checker.getTypeOfSymbolAtLocation(scope.symbol, scope) : checker.getDeclaredTypeOfSymbol(scope.symbol);
+            const props = checker.getPropertiesOfType(type);
+            functionNameText = getUniqueName(n => props.every(p => p.name !== n));

```

Comment:
>every [](start = 56, length = 5)

This appears to be quadratic in the number of un-renamed extracted functions.  That's not really an important case, but a set might still be worth considering.